source/common/io/io_uring_worker_impl.cc

```diff
@@ -0,0 +1,240 @@
+#include "source/common/io/io_uring_worker_impl.h"
+
+#include <sys/socket.h>
+
+#include "io_uring_impl.h"
+
+namespace Envoy {
+namespace Io {
+
+IoUringSocketEntry::IoUringSocketEntry(os_fd_t fd, IoUringWorkerImpl& parent)
+    : fd_(fd), parent_(parent) {}
+
+void IoUringSocketEntry::unlink() {
+  std::unique_ptr<IoUringSocketEntry> socket = parent_.removeSocket(*this);
+  parent_.dispatcher().deferredDelete(std::move(socket));
+}
+
+void IoUringSocketEntry::cleanup() {
+  parent_.removeInjectedCompletion(*this);
+  unlink();
+}
+
+void IoUringSocketEntry::injectCompletion(uint32_t type) {
+  // Avoid injected same type completion multiple times.
+  if (injected_completions_ & type) {
+    ENVOY_LOG(trace,
+              "ignore injected completion since there already has one, injected_completions_: {}, "
+              "type: {}",
+              injected_completions_, type);
+    return;
+  }
+  injected_completions_ |= type;
+  parent_.injectCompletion(*this, type, -EAGAIN);
+}
+
+IoUringWorkerImpl::IoUringWorkerImpl(uint32_t io_uring_size, bool use_submission_queue_polling,
+                                     Event::Dispatcher& dispatcher)
+    : IoUringWorkerImpl(std::make_unique<IoUringImpl>(io_uring_size, use_submission_queue_polling),
+                        dispatcher) {}
+
+IoUringWorkerImpl::IoUringWorkerImpl(std::unique_ptr<IoUring> io_uring_instance,
+                                     Event::Dispatcher& dispatcher)
+    : io_uring_instance_(std::move(io_uring_instance)), dispatcher_(dispatcher) {
+  const os_fd_t event_fd = io_uring_instance_->registerEventfd();
+  // We only care about the read event of Eventfd, since we only receive the
+  // event here.
+  file_event_ = dispatcher_.createFileEvent(
+      event_fd, [this](uint32_t) { onFileEvent(); }, Event::PlatformDefaultTriggerType,
+      Event::FileReadyType::Read);
+}
+
+IoUringWorkerImpl::~IoUringWorkerImpl() {
+  ENVOY_LOG(trace, "destruct io uring worker");
+  dispatcher_.clearDeferredDeleteList();
```

Comment:
Why do we need to clear the deferred deletion list to dtor this?

Reply:
Good question, When the IoUringWorker is gone, probably the whole thread is gone, I thought this is for cleanup all the derrred deleted items before the dispatcher is gone also. I saw we do the same thing for the listener. But I could understand this wrong.

I think what I missed here is when the IoUringWorkerImpl is gone, I should cleanup all the IoUringSockets here.